// @blink402/config/dist/logger.js
"use strict";
/**
 * Environment-aware logger for shared packages
 *
 * This logger works in both Node.js and browser environments.
 * It respects LOG_LEVEL environment variable and is silent in production by default.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
exports.createLogger = createLogger;
const LOG_LEVELS = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
    silent: 4,
};
class Logger {
    constructor(packageName = 'unknown') {
        this.packageName = packageName;
        // Determine log level from environment
        const envLevel = (typeof process !== 'undefined'
            ? process.env?.LOG_LEVEL || process.env?.NODE_ENV
            : 'production');
        // Default levels based on environment
        if (envLevel === 'production') {
            this.level = 'error'; // Only errors in production
        }
        else if (envLevel === 'test') {
            this.level = 'silent'; // Silent during tests
        }
        else {
            this.level = 'info'; // Info and above in development
        }
        // Allow explicit log level override
        if (LOG_LEVELS[envLevel] !== undefined) {
            this.level = envLevel;
        }
    }
    shouldLog(level) {
        return LOG_LEVELS[level] >= LOG_LEVELS[this.level];
    }
    formatMessage(level, message, meta) {
        const timestamp = new Date().toISOString();
        const prefix = `[${timestamp}] [${this.packageName}] ${level.toUpperCase()}`;
        if (meta && Object.keys(meta).length > 0) {
            return `${prefix}: ${message} ${JSON.stringify(meta)}`;
        }
        return `${prefix}: ${message}`;
    }
    debug(message, meta) {
        if (this.shouldLog('debug')) {
            console.debug(this.formatMessage('debug', message, meta));
        }
    }
    info(message, meta) {
        if (this.shouldLog('info')) {
            console.info(this.formatMessage('info', message, meta));
        }
    }
    warn(message, meta) {
        if (this.shouldLog('warn')) {
            console.warn(this.formatMessage('warn', message, meta));
        }
    }
    error(message, error, meta) {
        if (this.shouldLog('error')) {
            const errorMeta = error instanceof Error
                ? {
                    ...meta,
                    error: error.message,
                    stack: error.stack
                }
                : { ...meta, error: String(error) };
            console.error(this.formatMessage('error', message, errorMeta));
        }
    }
    /**
     * Set the log level dynamically
     */
    setLevel(level) {
        this.level = level;
    }
    /**
     * Get current log level
     */
    getLevel() {
        return this.level;
    }
}
/**
 * Create a logger instance for a specific package
 */
function createLogger(packageName) {
    return new Logger(packageName);
}
/**
 * Default logger instance
 */
exports.logger = new Logger('blink402');
